src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -145,27 +140,87 @@ internal abstract partial class AbstractProject : ForegroundThreadAffinitizedObj
             if (visualStudioWorkspaceOpt != null)
             {
                 this.EditAndContinueImplOpt = new VsENCRebuildableProjectImpl(this);
+                this.MetadataService = visualStudioWorkspaceOpt.Services.GetService<IMetadataService>();
             }
         }
 
         internal IServiceProvider ServiceProvider { get; }
-        
+
         /// <summary>
         /// Indicates whether this project is a website type.
         /// </summary>
         public bool IsWebSite { get; protected set; }
 
+        #region ObjOutputPath
+        private string _objOutputPath;
+
         /// <summary>
         /// A full path to the project obj output binary, or null if the project doesn't have an obj output binary.
         /// </summary>
-        internal string TryGetObjOutputPath() => _objOutputPathOpt;
+        internal string ObjOutputPath
+        {
+            get
+            {
+                lock (_gate)
+                {
+                    return _objOutputPath;
+                }
+            }
+            private set
+            {
+                lock (_gate)
+                {
+                    _objOutputPath = value;
+                }
+            }
+        }
+        #endregion
 
+        #region BinOutputPath
+        private string _binOutputPath;
+        
         /// <summary>
         /// A full path to the project bin output binary, or null if the project doesn't have an bin output binary.
         /// </summary>
-        internal string TryGetBinOutputPath() => _binOutputPathOpt;
+        internal string BinOutputPath
+        {
+            get
+            {
+                lock (_gate)
+                {
+                    return _binOutputPath;
+                }
+            }
+            private set
+            {
+                lock (_gate)
+                {
+                    _binOutputPath = value;
+                }
+            }
+        }
+        #endregion
 
-        public IRuleSetFile RuleSetFile { get; private set; }
+        #region RuleSetFile
+        private IRuleSetFile _ruleSetFile;
```

Comment:
Looks like you added comments about threading policy to all existing fields, but no comments for new fields.  Please make sure all fields have their threading policy addressed.